zerver/lib/actions.py

```diff
@@ -5282,6 +5294,32 @@ def do_change_user_setting(
             active_user_ids(user_profile.realm_id),
         )
 
+    if setting_name == "email_address_visibility":
+        person = dict(
+            user_id=user_profile.id,
+            email_address_visibility=user_profile.email_address_visibility,
+        )
+        send_event(
+            user_profile.realm,
+            dict(type="realm_user", op="update", person=person),
+            active_user_ids(user_profile.realm_id),
+        )
+
+        if UserProfile.EMAIL_ADDRESS_VISIBILITY_EVERYONE not in [old_value, setting_value]:
+            # We use real email addresses on UserProfile.email only if
+            # EMAIL_ADDRESS_VISIBILITY_EVERYONE is configured, so
+            # changes between values that will not require changing
+            # that field, so we can save work and return here.
+            return
+
+        user_profile.email = get_display_email_address(user_profile)
+        send_user_email_update_event(user_profile)
+        user_profile.save(update_fields=["email"])
+
+        flush_user_profile(sender=UserProfile, instance=user_profile)
+
+        notify_avatar_url_change(user_profile)
```

Comment:
I think there's a subtle issue here, where if a user changes their email address visibility to be more restrictive, we'd do an event that deletes the `delivery_email` values from others' clients' memory.  This may be a reason to send an `email_address_visibility` event even if we don't add those to user objects in general.

Reply:
I suppose one answer would be to send a `update_user` event with `delivery_email=null` as the event format to the set of users who lost access?